### PR TITLE
Sidebar top CPU time & allocated memory

### PR DESCRIPTION
**Pull request by ionutursuleanu**
Hi @trask 

I added the option to sort by total CPU time & allocated memory on the sidebar. Only the embedded collector will show percentages for old data. Would you consider include this into glowroot?
image

### CODE REVIEW

**Comment by trask** [comment]
Hey @ionutursuleanu, yes, this looks nice. Can I ask, what do you mean by

> Only the embedded collector will show percentages for old data.

**Comment by ionutursuleanu** [comment]
> Hey @ionutursuleanu, yes, this looks nice. Can I ask, what do you mean by
> 
> > Only the embedded collector will show percentages for old data.

By old data I mean what you see in glowroot before a version update. 

For the embedded collector I just used existing values from aggregate_tn_rollup_:
- total_cpu_nanos = sum(main_thread_total_cpu_nanos) + sum(aux_thread_total_cpu_nanos)
- total_allocated_bytes = sum(main_thread_total_allocated_bytes) + sum(aux_thread_total_allocated_bytes)

With the central collector was a bit more complicated due to cassandra tables structure. I had to add 2 new columns(total_cpu_nanos & total_allocated_bytes) into aggregate_tt_summary_rollup_ & aggregate_tn_summary_rollup_ Existing rows in those tables will have null as default value. Only new rows will be inserted will actual cpu time / allocated bytes. This is were I'm not sure if I can populate those 2 columns will something calculated from other tables.

**Comment by trask** [comment]
> By old data I mean what you see in glowroot before a version update.

Makes sense, that's fine.

File: agent/embedded/src/main/java/org/glowroot/agent/embedded/repo/AggregateDao.java [hunk 1]
```diff
@@ -740,14 +756,22 @@ public void bind(PreparedStatement preparedStatement) throws SQLException {
...
+            case TOTAL_CPU_TIME:
+                return "sum(main_thread_total_cpu_nanos) + sum(aux_thread_total_cpu_nanos) desc";
+            case AVERAGE_CPU_TIME:
+                return "(sum(main_thread_total_cpu_nanos) + sum(aux_thread_total_cpu_nanos)) / sum(transaction_count) desc";
+            case TOTAL_ALLOCATED_MEMORY:
+                return "sum(main_thread_total_allocated_bytes) + sum(aux_thread_total_allocated_bytes) desc";
+            case AVERAGE_ALLOCATED_MEMORY:
+                return "(sum(main_thread_total_allocated_bytes) + sum(aux_thread_total_allocated_bytes)) / sum(transaction_count) desc";
+            default:
+                throw new AssertionError("Unexpected sort order: " + sortOrder);
```
Comment (bmarwell) [inline agent/embedded/src/main/java/org/glowroot/agent/embedded/repo/AggregateDao.java:774]: Is this whitespace change necessary?

File: central/src/main/java/org/glowroot/central/repo/AggregateDaoImpl.java [hunk 2]
```diff
@@ -2079,9 +2091,46 @@ private static String mergeRowIntoSummaries(Row row, Map<String, MutableSummary>
...
         summary.totalDurationNanos += row.getDouble(i++);
         summary.transactionCount += row.getLong(i++);
+        summary.totalCpuNanos += row.getDouble(i++);
+        summary.totalAllocatedBytes += row.getDouble(i++);
         return transactionName;
     }
 
+    private static void bindAggregateForSummary(BoundStatement boundStatement, Aggregate aggregate,
+            int startIndex, TTL adjustedTTL) throws IOException {
+        int i = startIndex;
```
Comment (bmarwell) [inline central/src/main/java/org/glowroot/central/repo/AggregateDaoImpl.java:2101]: Although sibling methods do not have this and it is a private method, a short Javadoc would be really nice:
When is an `IOException` thrown?
What are valid parameter ranges?
What does the method do (short summary)?

File: common/src/main/java/org/glowroot/common/model/TransactionNameSummaryCollector.java [hunk 3]
```diff
@@ -94,19 +132,27 @@ public long getLastCaptureTime() {
...
+        case TOTAL_CPU_TIME:
+            return orderingByTotalCpuTimeDesc.immutableSortedCopy(transactionNameSummaries);
+        case AVERAGE_CPU_TIME:
+            return orderingByAverageCpuTimeDesc.immutableSortedCopy(transactionNameSummaries);
+        case TOTAL_ALLOCATED_MEMORY:
+            return orderingByTotalAllocatedMemoryDesc.immutableSortedCopy(transactionNameSummaries);
+        case AVERAGE_ALLOCATED_MEMORY:
+            return orderingByAverageAllocatedMemoryDesc.immutableSortedCopy(transactionNameSummaries);
+        default:
+            throw new AssertionError("Unexpected sort order: " + sortOrder);
```
Comment (bmarwell) [inline common/src/main/java/org/glowroot/common/model/TransactionNameSummaryCollector.java:150]: Same. Why change whitespace?

File: ui/app/scripts/routes.js [hunk 4]
```diff
@@ -177,9 +177,13 @@ glowroot.config([
       resolve: {
         summarySortOrders: function () {
           return {
-            'total-time': 'By percent of total time',
+            'total-time': 'By total time (%)',
```
Comment (bmarwell) [inline ui/app/scripts/routes.js:180]: Bravo 👏 for spotting this! :)

**Review by bmarwell — CHANGES_REQUESTED** [review]
Whitespace nits

**Comment by gonchik** [comment]
@trask  Do you have a plan to merge that PR ? 
What exactly need to be done to be merged?
I will be happy to start use as new release on my installations.

**Review by trask — APPROVED** [review]
@ionutursuleanu thx, everything looks great!

the only thing I'm still thinking about is that the dropdown options are a bit overwhelming, especially the difference between "total % and "average" is confusing for the different options (which was a problem before this PR too).

if I can't think of any better organization/display, will merge with current option list

**Comment by gonchik** [comment]
@trask  Awesome! Thank you!

**Comment by lfen21** [comment]
it's work very well. Thanks

**Pull request merged**